src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,300 @@
+            if (expirationTime < DateTime.UtcNow)
+                return false;
+
+            var newCacheItem = new LRUCacheItem<TKey, TValue>(key, value, expirationTime);
+            // just need to update value and move it to the top
+            if (_map.ContainsKey(key))
+            {
+                // make sure node hasn't been removed by a different thread
+                if (_map.TryGetValue(key, out var cacheItem))
+                {
+                    _eventQueue.Add(() =>
+                    {
+                        _doubleLinkedList.Remove(cacheItem);
+                        _eventQueue.Add(() => _doubleLinkedList.AddFirst(newCacheItem));
+
+                    });
+                }
+                else
+                {
+                    _eventQueue.Add(() => _doubleLinkedList.AddFirst(newCacheItem));
+                }
+
+                return true;
+            }
+            else
+            {
+                // if cache is full, then remove the least recently used node
+                if (_map.Count >= _capacity)
+                {                
+                    _eventQueue.Add(() =>
+                    {
+                        var lru = _doubleLinkedList.Last;
+                        _map.TryRemove(lru.Value.Key, out _);
+                        _doubleLinkedList.Remove(lru);
+                    });
+                }
+
+                // add the new node
+                _eventQueue.Add(() => _doubleLinkedList.AddFirst(newCacheItem));
+                _map[key] = newCacheItem;
+
+                return true;
+            }
+        }
+
+        /// Each time a node gets accessed, it gets moved to the beginning (head) of the list.
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (!_map.ContainsKey(key))
+            {
+                value = default;
+                return false;
+            }
+
+            // make sure node hasn't been removed by a different thread
+            if (_map.TryGetValue(key, out var cacheItem))
+                _eventQueue.Add(() =>
+                {
+                    _doubleLinkedList.Remove(cacheItem);
+                    _doubleLinkedList.AddFirst(cacheItem);
+                });
+
+            value = cacheItem != null ? cacheItem.Value : default;
+            return cacheItem != null;
+        }
+
+        /// Removes a particular key from the cache.
+        public bool TryRemove(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (!_map.ContainsKey(key))
+            {
+                value = default;
+                return false;
+            }
+
+            // check to make sure node wasn't removed by a different thread
+            if (!_map.TryGetValue(key, out var cacheItem))
```

Comment:
Why is this necessary considering the check on ln 199 and the actual removal on ln 214.